File: apps/web/src/lib/tien-len/rules.ts
import type { Card, CombinationType, PlayedCombination } from '@/types';

interface CombinationResult {
  valid: boolean;
  type: CombinationType | null;
  rank: number | null;
  length?: number;
  message?: string;
}

const RANK_ORDER: number[] = [3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15];

function getRankValues(): Record<number, number> {
  return {
    3: 1, 4: 2, 5: 3, 6: 4, 7: 5, 8: 6, 9: 7, 10: 8,
    11: 9, 12: 10, 13: 11, 14: 12, 15: 13,
  };
}

export function validateCombination(cards: Card[]): CombinationResult {
  if (!cards || cards.length === 0) {
    return { valid: false, type: null, rank: null, message: 'Không có bài được chọn' };
  }

  const len = cards.length;

  // Single
  if (len === 1) {
    return {
      valid: true,
      type: 'single',
      rank: cards[0].rank,
    };
  }

  // Sort by rank for analysis
  const sorted = [...cards].sort((a, b) => a.rank - b.rank);
  const ranks = sorted.map(c => c.rank);
  const uniqueRanks = [...new Set(ranks)];
  const suits = cards.map(c => c.suit);
  const uniqueSuits = [...new Set(suits)];

  // Pair (2 cards, same rank)
  if (len === 2 && uniqueRanks.length === 1) {
    return {
      valid: true,
      type: 'pair',
      rank: uniqueRanks[0],
    };
  }

  // Triple (3 cards, same rank)
  if (len === 3 && uniqueRanks.length === 1) {
    return {
      valid: true,
      type: 'triple',
      rank: uniqueRanks[0],
    };
  }

  // Bomb (4 cards, same rank)
  if (len === 4 && uniqueRanks.length === 1) {
    return {
      valid: true,
      type: 'bomb',
      rank: uniqueRanks[0],
    };
  }

  // Straight (3+ consecutive cards, same suit)
  if (len >= 3 && uniqueSuits.length === 1) {
    const isConsecutive = checkConsecutive(ranks);
    if (isConsecutive) {
      // Cannot have 2 in straight (Ace-high only)
      if (ranks.includes(15)) {
        // Check if it's valid (not A-2-3)
        const aceIndex = ranks.indexOf(14);
        const deuceIndex = ranks.indexOf(15);
        if (aceIndex !== -1 && deuceIndex !== -1) {
          // A-2-3 is invalid
          const minRank = Math.min(...ranks);
          if (minRank === 3 && ranks.length === 3) {
            return { valid: false, type: null, rank: null, message: 'A-2-3 không hợp lệ trong Tiến Lên' };
          }
        }
      }
      return {
        valid: true,
        type: 'straight',
        rank: Math.max(...ranks), // Higher last card wins
        length: len,
      };
    }
  }

  // Full House (5 cards: 3 of a kind + 1 pair)
  if (len === 5) {
    const rankCounts = new Map<number, number>();
    for (const rank of ranks) {
      rankCounts.set(rank, (rankCounts.get(rank) || 0) + 1);
    }
    const counts = Array.from(rankCounts.values()).sort((a, b) => b - a);
    if (counts[0] === 3 && counts[1] === 2) {
      const tripleRank = Array.from(rankCounts.entries()).find(([, [, c]]) => c === 3)?.[0] || 0;
      return {
        valid: true,
        type: 'full_house',
        rank: tripleRank,
      };
    }
  }

  return { valid: false, type: null, rank: null, message: 'Bộ bài không hợp lệ' };
}

function checkConsecutive(ranks: number[]): boolean {
  const sorted = [...ranks].sort((a, b) => a - b);
  for (let i = 1; i < sorted.length; i++) {
    if (sorted[i] !== sorted[i - 1] + 1) {
      return false;
    }
  }
  return true;
}

export function canBeatCombination(
  newCards: Card[],
  lastPlay: PlayedCombination | null
): CombinationResult {
  const newResult = validateCombination(newCards);

  if (!newResult.valid) {
    return newResult;
  }

  if (!lastPlay) {
    return newResult;
  }

  // Bomb always beats non-bomb
  if (newResult.type === 'bomb' && lastPlay.combinationType !== 'bomb') {
    return newResult;
  }

  // Non-bomb cannot beat bomb
  if (newResult.type !== 'bomb' && lastPlay.combinationType === 'bomb') {
    return {
      valid: false,
      type: newResult.type,
      rank: newResult.rank,
      message: 'Phải đánh tứ quý để chặn tứ quý',
    };
  }

  // Types must match
  if (newResult.type !== lastPlay.combinationType) {
    return {
      valid: false,
      type: newResult.type,
      rank: newResult.rank,
      message: 'Phải đánh cùng loại bộ bài',
    };
  }

  // For straight, longer wins, then higher rank
  if (newResult.type === 'straight') {
    if ((newResult.length || 0) > (lastPlay.cards.length)) {
      return newResult;
    }
    if ((newResult.length || 0) < (lastPlay.cards.length)) {
      return {
        valid: false,
        type: newResult.type,
        rank: newResult.rank,
        message: 'Sảnh dài hơn thắng sảnh ngắn hơn',
      };
    }
  }

  // Rank must be higher
  if ((newResult.rank || 0) <= (lastPlayCardsMaxRank(lastPlay))) {
    return {
      valid: false,
      type: newResult.type,
      rank: newResult.rank,
      message: 'Bộ bài không đủ mạnh để chặn',
    };
  }

  return newResult;
}

function lastPlayCardsMaxRank(play: PlayedCombination): number {
  if (play.combinationType === 'straight') {
    return Math.max(...play.cards.map(c => c.rank));
  }
  return play.cards[0]?.rank || 0;
}

export function getComboRankValue(rank: number): number {
  const rankValues = getRankValues();
  return rankValues[rank] || 0;
}

export function compareRanks(rank1: number, rank2: number): number {
  return getComboRankValue(rank1) - getComboRankValue(rank2);
}

export function isValidFirstPlay(cards: Card[], hasThreeOfDiamonds: boolean): boolean {
  if (hasThreeOfDiamonds) {
    return cards.some(c => c.suit === 'diamonds' && c.rank === 3);
  }
  return false;
}

File: apps/api/src/game_logic/rules.py
from typing import List, Optional, Tuple
from src.game_logic.deck import Card, Suit

CombinationType = str  # "single", "pair", "triple", "straight", "full_house", "bomb"

def validate_combination(cards: List[Card]) -> Tuple[bool, Optional[CombinationType], Optional[int], Optional[str]]:
    if not cards:
        return False, None, None, "Không có bài được chọn"

    if len(cards) == 1:
        return True, "single", cards[0].rank, None

    # Sort by rank
    sorted_cards = sorted(cards, key=lambda c: c.rank)
    ranks = [c.rank for c in sorted_cards]
    suits = [c.suit for c in cards]
    unique_ranks = list(set(ranks))
    unique_suits = list(set(suits))

    # Pair (2 cards, same rank)
    if len(cards) == 2 and len(unique_ranks) == 1:
        return True, "pair", unique_ranks[0], None

    # Triple (3 cards, same rank)
    if len(cards) == 3 and len(unique_ranks) == 1:
        return True, "triple", unique_ranks[0], None

    # Bomb (4 cards, same rank)
    if len(cards) == 4 and len(unique_ranks) == 1:
        return True, "bomb", unique_ranks[0], None

    # Straight (3+ consecutive cards, same suit)
    if len(cards) >= 3 and len(unique_suits) == 1:
        is_consecutive = check_consecutive(ranks)
        if is_consecutive:
            # Check no 2 in straight (A-2-3 invalid)
            if 15 in ranks and 3 in ranks and len(cards) == 3:
                return False, None, None, "A-2-3 không hợp lệ"
            return True, "straight", max(ranks), None

    # Full House (3 + 2)
    if len(cards) == 5:
        rank_counts = {}
        for r in ranks:
            rank_counts[r] = rank_counts.get(r, 0) + 1
        counts = sorted(rank_counts.values(), reverse=True)
        if counts[0] == 3 and counts[1] == 2:
            triple_rank = [r for r, c in rank_counts.items() if c == 3][0]
            return True, "full_house", triple_rank, None

    return False, None, None, "Bộ bài không hợp lệ"

def check_consecutive(ranks: List[int]) -> bool:
    sorted_ranks = sorted(ranks)
    for i in range(1, len(sorted_ranks)):
        if sorted_ranks[i] != sorted_ranks[i-1] + 1:
            return False
    return True

def can_beat(new_cards: List[Card], last_combo_type: str, last_rank: int, last_length: int) -> Tuple[bool, Optional[str]]:
    is_valid, combo_type, rank, error = validate_combination(new_cards)
    if not is_valid:
        return False, error

    # Bomb beats everything
    if combo_type == "bomb" and last_combo_type != "bomb":
        return True, None

    # Non-bomb cannot beat bomb
    if combo_type != "bomb" and last_combo_type == "bomb":
        return False, "Phải đánh tứ quý để chặn tứ quý"

    # Must be same type
    if combo_type != last_combo_type:
        return False, "Phải đánh cùng loại bộ bài"

    # For straight, longer wins
    if combo_type == "straight":
        if len(new_cards) > last_length:
            return True, None
        if len(new_cards) < last_length:
            return False, "Sảnh dài hơn thắng sảnh ngắn hơn"

    # Rank must be higher
    rank_value = get_rank_value(rank)
    last_rank_value = get_rank_value(last_rank)
    if rank_value <= last_rank_value:
        return False, "Bộ bài không đủ mạnh để chặn"

    return True, None

def get_rank_value(rank: int) -> int:
    # 3=1, 4=2, ..., K=11, A=12, 2=13
    return rank - 2  # So 3 becomes 1

def must_play_three_diamonds(cards: List[Card], has_started: bool) -> bool:
    if has_started:
        return False
    return any(c.suit == Suit.DIAMONDS and c.rank == 3 for c in cards)
